### PR TITLE
NIFI-106: Expose processors' counters in Stats History

### PR DESCRIPTION
**Pull request by markap14**
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by mcgilman** [comment]
@markap14 - Counters are currently protected in access policies using a global resource. This is due to the fact that there are instance based Counters and a corresponding type based Counter. The type base Counters aggregate the values from all instances of that type reporting that Counter. Because type based Counters lose the association to the underlying instance, we are currently requiring a global permission. 

Access to stats and status history is available to any user with permissions to the UI (protected in access policies with the /flow resource). Because of this, I think we need to filter only allowable Counters into the status history. We will need to decide what is allowable in this context.

**Comment by markap14** [comment]
@mcgilman great point! When I looked through the code, I saw the section of code that was handling permissions around the Status History but didn't look close enough. I had assumed that if the user doesn't have read permissions on the processor, that the stats would be made unavailable. However, looking again, I see that what we do is allow access to the stats but just blank out the name of the Processor.

What probably makes sense, then, is to filter out any stats that come from counters unless the user does have read permissions to the Processor. I.e., if User A has no read permissions to a processor, its stats would return just the way they are now, with just the pre-defined Bytes In, Bytes Out, FlowFiles In, FlowFiles out, etc. But if the user does have read permissions, then they will also see all stats for counters emitted by that processor (but not type-specific counters, just instance-specific counters). Thoughts?

**Comment by gresockj** [comment]
If you think it's appropriate to include in this ticket, you could also update the SiteToSiteStatusReportingTask to include the counters in the serialized JSON.

**Comment by markap14** [comment]
@mcgilman I pushed a new commit that addresses your concern, such that counters are included in the stats only if the user is authorized to read the processor that the counters belong to. Of note, all nodes in the cluster must also agree that counters should be shown (in case they differ in their policies).

**Comment by mcgilman** [comment]
@markap14 this sounds like a good approach. Will review...

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/http/endpoints/StatusHistoryEndpointMerger.java [hunk 1]
```diff
@@ -109,13 +119,49 @@ public NodeResponse merge(URI uri, String method, Set<NodeResponse> successfulRe
                 noReadPermissionsComponentDetails = nodeStatus.getComponentDetails();
             }
 
+            if (!nodeStatus.isIncludeCounters()) {
```
Comment (mcgilman) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/http/endpoints/StatusHistoryEndpointMerger.java]: I'm not sure we need to add a new field to the `nodeStatus` as the read permission is already present in the corresponding `nodeResponseEntity`.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/FlowController.java [hunk 2]
```diff
@@ -2867,6 +2867,7 @@ private ProcessorStatus getProcessorStatus(final RepositoryStatusReport report,
             status.setFlowFilesSent(entry.getFlowFilesSent());
             status.setBytesSent(entry.getBytesSent());
             status.setFlowFilesRemoved(entry.getFlowFilesRemoved());
+            status.setCounters(entry.getCounters());
```
Comment (mcgilman) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/FlowController.java]: This should be done conditionally based on `isProcessorAuthorized`. When captured for status history purposes that `Predicate` will always result in `true`.

File: nifi-framework-api/src/main/java/org/apache/nifi/controller/status/history/StatusHistory.java [hunk 3]
```diff
@@ -41,4 +41,9 @@
      * @return List of snapshots for a given component
      */
     List<StatusSnapshot> getStatusSnapshots();
+
+    /**
+     * @return <code>true</code> if counter values are included in the Status History
+     */
+    boolean isIncludeCounters();
```
Comment (mcgilman) [inline nifi-framework-api/src/main/java/org/apache/nifi/controller/status/history/StatusHistory.java]: If we're able to remove the flag from the `StatusHistoryDTO`, we may also be able to remove this one.

File: nifi-api/src/main/java/org/apache/nifi/controller/status/ProcessorStatus.java [hunk 4]
```diff
@@ -234,6 +245,7 @@ public ProcessorStatus clone() {
         clonedObj.flowFilesRemoved = flowFilesRemoved;
         clonedObj.runStatus = runStatus;
         clonedObj.type = type;
+        clonedObj.counters = new HashMap<>(counters);
```
Comment (mcgilman) [inline nifi-api/src/main/java/org/apache/nifi/controller/status/ProcessorStatus.java]: May need to protect against NPE when `counters` is null.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/http/endpoints/StatusHistoryEndpointMerger.java (hunk 1 above)
Comment (markap14) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/http/endpoints/StatusHistoryEndpointMerger.java]: Good call. I didn't realize that was part of the entity.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/FlowController.java (hunk 2 above)
Comment (markap14) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/FlowController.java]: Agreed.

File: nifi-framework-api/src/main/java/org/apache/nifi/controller/status/history/StatusHistory.java (hunk 3 above)
Comment (markap14) [inline nifi-framework-api/src/main/java/org/apache/nifi/controller/status/history/StatusHistory.java]: Noted.

File: nifi-api/src/main/java/org/apache/nifi/controller/status/ProcessorStatus.java (hunk 4 above)
Comment (markap14) [inline nifi-api/src/main/java/org/apache/nifi/controller/status/ProcessorStatus.java]: Good catch. Will add a new commit shortly.

**Comment by markap14** [comment]
@mcgilman ok i pushed a new commit. I also ran into another bug that occurred when a counter is present in some of the 'aggregate snapshot' fields but not all. Can you give it a review when you have a chance? THanks!

**Comment by mcgilman** [comment]
Thanks @markap14! This has been merged to master.

**Pull request closed without merging**